roles/postgresql-function/tasks/main.yml

```diff
@@ -0,0 +1,14 @@
+---
+
+- name: Copy function file
+  ansible.builtin.copy:
+    src: "{{ role_path }}/files/pgbouncer_function.psql"
+    dest: /tmp
+
+- name: Create function for pgbouncer to query users
+  become: true
+  become_user: postgres
+  shell: |
```

Comment:
Note that the query is run inside the target database. So if a function is used, it needs to be installed into each database.

thus, it is necessary to execute code for all databases in the cluster, including the template1 database, so that this function is available in all new databases.

Reply:
Are you sure? Because every database I create after the function it also inherits the function:
```bash
root@patroni2:~# psql -h 192.168.100.105 -p 5000 -U postgres -d test2
Password for user postgres:
psql (15.3 (Ubuntu 15.3-1.pgdg22.04+1))
Type "help" for help.

test2=# \df+
                                                                                                              List of functions
 Schema |    Name     |         Result data type         | Argument data types | Type | Volatility | Parallel |  Owner   | Security | Access privileges | Language |                        Source code                        | Description
--------+-------------+----------------------------------+---------------------+------+------------+----------+----------+----------+-------------------+----------+-----------------------------------------------------------+-------------
 public | user_search | TABLE(usename name, passwd text) | uname text          | func | volatile   | unsafe   | postgres | definer  |                   | sql      |                                                          +|
        |             |                                  |                     |      |            |          |          |          |                   |          |   SELECT usename, passwd FROM pg_shadow WHERE usename=$1;+|
        |             |                                  |                     |      |            |          |          |          |                   |          |                                                           |
(1 row)
```

Reply:
I am sure that the function should also be created in the template1 database

Reply:
Thats strange, because I didn't explicitly created the function in template1:

```bash
root@patroni2:~# psql -h 192.168.100.105 -p 5000 -U postgres -d template1
Password for user postgres:
psql (15.3 (Ubuntu 15.3-1.pgdg22.04+1))
Type "help" for help.

template1=# \df
                                  List of functions
 Schema |    Name     |         Result data type         | Argument data types | Type
--------+-------------+----------------------------------+---------------------+------
 public | user_search | TABLE(usename name, passwd text) | uname text          | func
(1 row)

template1=#
```

Reply:
by default, databases are created based on the template1 template (if another template is not specified during database creation). Therefore, in order to guarantee the inheritance of the function for all new databases, we must create it in the template1 database.

Reply:
OK, understood